src/apps/config/alarm_codec.lua

```diff
@@ -0,0 +1,195 @@
+-- Use of this source code is governed by the Apache 2.0 license; see COPYING.
+
+module(...,package.seeall)
+
+local S = require("syscall")
+local channel = require("apps.config.channel")
+local ffi = require("ffi")
+
+local alarm_names = { 'raise_alarm', 'clear_alarm' }
+local alarm_codes = {}
+for i, name in ipairs(alarm_names) do alarm_codes[name] = i end
+
+local alarms = {}
+
+function alarms.raise_alarm (codec, resource, alarm_type_id, alarm_type_qualifier,
+   perceived_severity, alarm_text)
+
+   local resource = codec:string(resource)
+   local alarm_type_id = codec:string(alarm_type_id)
+   local alarm_type_qualifier = codec:string(alarm_type_qualifier)
+
+   local perceived_severity = codec:string(perceived_severity)
+   local alarm_text = codec:string(alarm_text)
+
+   return codec:finish(resource, alarm_type_id, alarm_type_qualifier,
+                       perceived_severity, alarm_text)
+end
+function alarms.clear_alarm (codec, resource, alarm_type_id, alarm_type_qualifier)
+   local resource = codec:string(resource)
+   local alarm_type_id = codec:string(alarm_type_id)
+   local alarm_type_qualifier = codec:string(alarm_type_qualifier)
+
+   return codec:finish(resource, alarm_type_id, alarm_type_qualifier)
+end
+
+local function encoder()
+   local encoder = { out = {} }
```

Comment:
I think that since these messages come from the data plane, in the future we're going to want to use a static byte buffer large enough to hold any message, and then make e.g. `encoder:uint32()` just write into that buffer and advance the write pointer.  This avoids allocation and thus GC pressure.  But that can be for a followup.